### PR TITLE
Standardize specific User Preferences

### PR DESCRIPTION
**Pull request by Baggerone**
I created graphql enums for the language and weighUnit values. 

for example:
```
mutation { user: updateUser(input:{id: "6fc44091-c8bf-485a-9d46-a28b0f686af6", preferences: {weightUnit: KILOGRAMS}}) {id nickname preferences {language, timeZone, weightUnit}} }
```

### CODE REVIEW

File: application/gqlgen/gqlgenutils.go [hunk 1]
```diff
@@ -37,13 +38,34 @@ func convertGqlLocationInputToDBLocation(input LocationInput) models.Location {
...
+	if input.Language != nil {
+		lang := strings.ToLower(fmt.Sprintf("%v", *input.Language))
+		if !domain.IsLanguageAllowed(lang) {
+			return models.StandardPreferences{}, errors.New("user preference language not allowed ... " + lang)
+		}
+		stPrefs.Language = lang
+	}
+
+	if input.TimeZone != nil {
+		stPrefs.TimeZone = strings.ToLower(*input.TimeZone)
```
Comment (fillup) [inline application/gqlgen/gqlgenutils.go]: We'll need some validation of the timezone

File: application/models/user_test.go [hunk 2]
```diff
@@ -1079,109 +1079,146 @@ func (ms *ModelSuite) TestUser_WantsPostNotification() {
...
+		{
+			name: "Start with none then add one",
+			user: f.Users[1],
+			SPrefs: StandardPreferences{
+				Language: domain.UserPreferenceLanguageFrench,
+			},
+			want: StandardPreferences{
+				Language: domain.UserPreferenceLanguageFrench,
+			},
+		},
```
Comment (Schparky) [inline application/models/user_test.go]: Seems there could be other possibilities. For instance, will it work correctly if a user has one preference and another is added? Just to make certain that existing one doesn't get changed.

File: application/domain/domain_test.go [hunk 3]
```diff
@@ -454,3 +454,17 @@ func (ts *TestSuite) TestIsWeightUnitAllowed() {
...
+	got := IsTimeZoneAllowed(zone)
+	ts.True(got, zone+" should be an allowed time zone")
+
+	zone = "Etc/GMT+2"
+	got = IsTimeZoneAllowed(zone)
+	ts.True(got, zone+" should be an allowed time zone")
+
+	zone = "badzone"
+	got = IsLanguageAllowed(zone)
+	ts.False(got, zone+" should not be an allowed language")
```
Comment (Schparky) [inline application/domain/domain_test.go]: Is this a copy-paste error? Should be `IsTimeZoneAllowed(zone)` and `should not be an allowed time zone`

File: application/gqlgen/gqlgenutils.go (hunk 1 above)
Comment (Baggerone) [reply application/gqlgen/gqlgenutils.go]: Done

File: application/gqlgen/user_test.go [hunk 4]
```diff
@@ -176,7 +177,8 @@ func (gs *GqlgenSuite) TestUserQuery() {
 
 	var resp UserResponse
 
-	allFields := `{ id email nickname adminRole photoURL preferences {key}  posts (role: CREATEDBY) {id} organizations {id}
+	allFields := `{ id email nickname adminRole photoURL preferences {language weightUnit}
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply application/gqlgen/user_test.go]: Done

File: application/domain/domain_test.go (hunk 3 above)
Comment (Baggerone) [reply application/domain/domain_test.go]: Yep. Good catch.

File: application/domain/domain_test.go [hunk 5]
```diff
@@ -2,15 +2,32 @@ package domain
...
+// Test_GqlgenSuite runs the GqlgenSuite test suite
+func Test_TestSuite(t *testing.T) {
+	model := suite.NewModel()
+
+	gs := &TestSuite{
+		Model: model,
+	}
+	suite.Run(t, gs)
+}
+
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply application/domain/domain_test.go]: Good. Thanks.

File: application/models/user.go [hunk 6]
```diff
@@ -528,35 +528,57 @@ func (u *User) WantsPostNotification(post Post) bool {
 	return true
 }
 
-// GetPreferences returns a slice of matching UserPreference records
-func (u *User) GetPreferences() (UserPreferences, error) {
+// GetPreferences returns an StandardPreferences struct
+func (u *User) GetPreferences() (StandardPreferences, error) {
 	uPrefs := UserPreferences{}
-	err := DB.Where("user_id = ?", u.ID).All(&uPrefs)
+	if err := DB.Where("user_id = ?", u.ID).All(&uPrefs); err != nil {
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply application/models/user.go]: Sounds good.

File: application/models/user_test.go (hunk 2 above)
Comment (Baggerone) [reply application/models/user_test.go]: That's what the previous test does.  The user has French and Kilograms and no Time Zone. The test changes French to English, leaves Kilograms as is and adds a Time Zone.

Is that what you mean?

Comment (Schparky) [reply application/models/user_test.go]: Yes, I think so. I admit I didn't study it that well.

File: application/models/models.go [hunk 7]
```diff
@@ -39,6 +44,11 @@ func init() {
 	if _, err = getRandomToken(); err != nil {
 		log.Fatal(fmt.Errorf("error using crypto/rand ... %v", err))
 	}
+
+	allowedUserPreferenceKeys, err = domain.GetStructTags("json", StandardPreferences{})
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply application/models/models.go]: Thanks.

**Pull request merged**